File: codes/rust/480.滑动窗口中位数.rs
/*
 * @lc app=leetcode.cn id=480 lang=rust
 *
 * [480] 滑动窗口中位数
 */

// @lc code=start

use std::cmp::Reverse;
use std::collections::{BinaryHeap, HashMap};

struct DualHeap {
    // 大根堆，维护较小的一半元素
    small: BinaryHeap<i32>,
    // 小根堆，维护较大的一半元素
    large: BinaryHeap<Reverse<i32>>,
    // 哈希表，记录延迟删除的元素,key为元素，value为需要删除的次数
    delayed: HashMap<i32, i32>,
    // small和large当前包含的元素个数，需要扣除被【延迟删除】的元素
    small_size: usize,
    large_size: usize,
    // 元素个数大小
    k: usize,
}

impl DualHeap {
    fn new(k: usize) -> DualHeap {
        DualHeap {
            small: BinaryHeap::new(),
            large: BinaryHeap::new(),
            delayed: HashMap::new(),
            small_size: 0,
            large_size: 0,
            k,
        }
    }

    // 获取中位数
    fn get_median(&self) -> f64 {
        if self.k & 1 == 1 {
            return self.small.peek().unwrap().to_owned() as f64;
        }
        let left_val = self.small.peek().unwrap().to_owned() as f64;
        let right_val = self.large.peek().unwrap().0.to_owned() as f64;
        (left_val + right_val) / 2.0
    }

    // 插入数据
    fn insert(&mut self, num: i32) {
        if self.small.is_empty() || num <= self.small.peek().unwrap().to_owned() {
            self.small.push(num);
            self.small_size += 1;
        } else {
            self.large.push(Reverse(num));
            self.large_size += 1;
        }
        self.make_balance()
    }

    // 擦除数据
    fn erase(&mut self, num: i32) {
        let mut count = 0;
        if let Some(cnt) = self.delayed.get(&num) {
            count = cnt.to_owned();
        }
        count += 1;
        self.delayed.insert(num, count);
        let small_top_val = self.small.peek().unwrap().to_owned();
        if num <= small_top_val {
            self.small_size -= 1;
            if num == small_top_val {
                self.prune_small();
            }
        } else {
            self.large_size -= 1;
            if num == self.large.peek().unwrap().to_owned().0 {
                self.prune_large();
            }
        }
        self.make_balance()
    }

    // 不断地弹出大根heap的堆顶元素，并且更新哈希表
    fn prune_small(&mut self) {
        let mut heap = &mut self.small;
        while !heap.is_empty() {
            let top_num = heap.peek().unwrap().to_owned();
            if self.delayed.contains_key(&top_num) {
                let mut count = self.delayed.get(&top_num).unwrap().to_owned();
                count -= 1;
                if count == 0 {
                    self.delayed.remove(&top_num);
                } else {
                    self.delayed.insert(top_num, count);
                }
                heap.pop();
            } else {
                break;
            }
        }
    }

    // 不断地弹出小根heap的堆顶元素，并且更新哈希表
    fn prune_large(&mut self) {
        let mut heap = &mut self.large;
        while !heap.is_empty() {
            let top_num = heap.peek().unwrap().to_owned().0;
            if self.delayed.contains_key(&top_num) {
                let mut count = self.delayed.get(&top_num).unwrap().to_owned();
                count -= 1;
                if count == 0 {
                    self.delayed.remove(&top_num);
                } else {
                    self.delayed.insert(top_num, count);
                }
                heap.pop();
            } else {
                break;
            }
        }
    }

    // 调整small和large中的元素个数，使得二者的元素个数满足要求
    fn make_balance(&mut self) {
        if self.small_size > self.large_size + 1 {
            // small比large元素多两个
            let small_top_val = self.small.pop().unwrap();
            self.small_size -= 1;
            self.large.push(Reverse(small_top_val));
            self.large_size += 1;
            // small堆顶元素被移除，需要进行prune
            self.prune_small();
        } else if self.small_size < self.large_size {
            // large比small元素多一个
            let large_top_val = self.large.pop().unwrap();
            self.large_size -= 1;
            self.small.push(large_top_val.0);
            self.small_size += 1;
            // large堆顶元素被移除，需要进行prune
            self.prune_large();
        }
    }
}

impl Solution {
    pub fn median_sliding_window(nums: Vec<i32>, k: i32) -> Vec<f64> {
        let mut my_heap = DualHeap::new(k as usize);
        for i in 0..k as usize {
            my_heap.insert(nums[i])
        }
        let mut result = Vec::new();
        result.push(my_heap.get_median());
        for j in k as usize..nums.len() {
            // 先从窗口中擦除掉第一个元素
            my_heap.erase(nums[j - k as usize]);
            my_heap.insert(nums[j]);
            // 再获取中位数
            result.push(my_heap.get_median());
        }
        result
    }
}
// @lc code=end

File: codes/java/480.滑动窗口中位数.java
import java.util.PriorityQueue;
import java.util.HashMap;

// 定义一个双堆数据结构，其中维护一个大根堆和一个小根堆
class DualHeap {
	// 大根堆放置较小的一半元素
	private PriorityQueue<Integer> small;
	// 小根堆放置较大的一半元素
	private PriorityQueue<Integer> large;
	// 堆大小
	private int k;
	// 大根堆和小根堆元素个数
	private int smallSize, largeSize;
	// 延迟删除的元素，key为元素，value为需要删除的次数
	private HashMap<Integer, Integer> delayed;

	public DualHeap(int k) {
		this.small = new PriorityQueue<Integer>(new Comparator<Integer>() {
			public int compare(Integer num1, Integer num2) {
				return num2.compareTo(num1);
			}
		});
		this.large = new PriorityQueue<Integer>(new Comparator<Integer>() {
			public int compare(Integer num1, Integer num2) {
				return num1.compareTo(num2);
			}
		});
		this.k = k;
		this.smallSize = 0;
		this.largeSize = 0;
		this.delayed = new HashMap<Integer, Integer>();
	}

	// 获取中位数
	public double getMedian() {
		return (k & 1) == 1 ? small.peek() : ((double) small.peek() + large.peek()) / 2;
	}

	// 插入数据
	public void insert(int num) {
		if (small.isEmpty() || num <= small.peek()) {
			small.offer(num);
			++smallSize;
		} else {
			large.offer(num);
			++largeSize;
		}
		makeBalance();
	}

	// 删除数据
	public void erase(int num) {
		delayed.put(num, delayed.getOrDefault(num, 0) + 1);
		if (num <= small.peek()) {
			--smallSize;
			if (num == small.peek()) {
				prune(small);
			}
		} else {
			--largeSize;
			if (num == large.peek()) {
				prune(large);
			}
		}
		makeBalance();
	}

	// 将堆顶元素移除
	private void prune(PriorityQueue<Integer> heap) {
		while (!heap.isEmpty()) {
			int num = heap.peek();
			if (delayed.containsKey(num)) {
				delayed.put(num, delayed.get(num) - 1);
				if (delayed.get(num) == 0) {
					delayed.remove(num);
				}
				heap.poll();
			} else {
				break;
			}
		}
	}

	// 调整两个堆的大小
	private void makeBalance() {
		if (smallSize > largeSize + 1) {
			large.offer(small.poll());
			--smallSize;
			++largeSize;
			prune(small);
		} else if (smallSize < largeSize) {
			small.offer(large.poll());
			++smallSize;
			--largeSize;
			prune(large);
		}
	}
}

class Solution {
	public double[] medianSlidingWindow(int[] nums, int k) {
		DualHeap dh = new DualHeap(k);
		for (int i = 0; i < k; ++i) {
			dh.insert(nums[i]);
		}
		double[] ans = new double[nums.length - k + 1];
		ans[0] = dh.getMedian();
		for (int i = k; i < nums.length; ++i) {
			dh.insert(nums[i]);
			dh.erase(nums[i - k]);
			ans[i - k + 1] = dh.getMedian();
		}
		return ans;
	}
}